src/main/java/org/opensearch/ad/task/ADTaskManager.java

```diff
@@ -405,6 +517,235 @@ protected void forwardStaleRunningEntitiesToCoordinatingNode(
+            ADStatsRequest adStatsRequest = new ADStatsRequest(nodes);
+            adStatsRequest
+                .addAll(ImmutableSet.of(AD_USED_BATCH_TASK_SLOT_COUNT.getName(), AD_DETECTOR_ASSIGNED_BATCH_TASK_SLOT_COUNT.getName()));
+            client.execute(ADStatsNodesAction.INSTANCE, adStatsRequest, ActionListener.wrap(adStatsResponse -> {
+                int totalUsedTaskSlots = 0; // Total entity tasks running on worker nodes
+                int totalAssignedTaskSlots = 0; // Total assigned task slots on coordinating nodes
+                for (ADStatsNodeResponse response : adStatsResponse.getNodes()) {
+                    totalUsedTaskSlots += (int) response.getStatsMap().get(AD_USED_BATCH_TASK_SLOT_COUNT.getName());
+                    totalAssignedTaskSlots += (int) response.getStatsMap().get(AD_DETECTOR_ASSIGNED_BATCH_TASK_SLOT_COUNT.getName());
+                }
+                logger
+                    .info(
+                        "Current total used task slots is {}, total detector assigned task slots is {} when start historical "
+                            + "analysis for detector {}",
+                        totalUsedTaskSlots,
+                        totalAssignedTaskSlots,
+                        detectorId
+                    );
+                // In happy case, totalAssignedTaskSlots >= totalUsedTaskSlots. If some coordinating node left, then we can't
+                // get detector task slots cached on it, so it's possible that totalAssignedTaskSlots < totalUsedTaskSlots.
+                int currentUsedTaskSlots = Math.max(totalUsedTaskSlots, totalAssignedTaskSlots);
+                if (currentUsedTaskSlots >= maxAdTaskSlots) {
+                    wrappedActionListener.onFailure(new OpenSearchStatusException("No available task slot", RestStatus.BAD_REQUEST));
+                    return;
+                }
+                int availableAdTaskSlots = maxAdTaskSlots - currentUsedTaskSlots;
+                logger.info("Current available task slots is {} for historical analysis of detector {}", availableAdTaskSlots, detectorId);
+
+                if (ADTaskAction.SCALE_ENTITY_TASK_SLOTS == targetActionOfTaskSlotChecking) {
+                    forwardToCoordinatingNode(
+                        adTask,
+                        detector,
+                        detectionDateRange,
+                        user,
+                        targetActionOfTaskSlotChecking,
+                        transportService,
+                        wrappedActionListener,
+                        availableAdTaskSlots
+                    );
+                    return;
+                }
+
+                long dataStartTime = detectionDateRange.getStartTime().toEpochMilli();
+                long dataEndTime = detectionDateRange.getEndTime().toEpochMilli();
+
+                if (detector.isMultientityDetector()) {
+                    if (detector.isMultiCategoryDetector()) {
```

Comment:
May be helpful to add some comments on the different searches that happen based on if the detector is single entity vs. multi entity (single category) vs. multi entity (multi category)

Reply:
sure, will add